[trainer] remove `--model_parallel`

stas00 opened the pull request:
Per @sgugger's request removing `--model_parallel` in trainer, as it was never tested or made to work with the trainer.

We will get back to it in the future.

This PR doesn't introduce breaking changes, since  `--model_parallel` never worked (well other than in my MP PRs that have been parked for now, since they are very inefficient and we are looking for a better approach, rather than waste time on sorting those out).

@LysandreJik, @sgugger

LysandreJik reviewed (approved):
Indeed, LGTM! We should have been more attentive during the review, no harm done.

@sgugger for info, this was removed here: https://github.com/huggingface/transformers/commit/9f675b05d4770655afa90ef51333aec032021a6b#diff-ed55888e6665791fe92cc8fc0c499da54f4ace6738551cd9a2591881cda076deL245-L248

sgugger commented:
Thanks for putting it back. Since we're in a PR on this test alone, can we "fix" it to ignore the `args.model_parallel` argument? This argument will be removed/renamed (I'd prefer the first option as it's not useful) since peoples are confusing it with something that will enable `DataParallel`. The test can be replaced by `model.is_parallelizable and model.parallel` I believe, with the current API.

sgugger commented:
> i.e. MP works when setup up manually but doesn't work in trainer.
> As we discovered last night current trainer doesn't work at all with --model_parallel - see #9211 (comment) there is no activation of that parallel mode - nobody calls model.parallelize() so it's very broken

That's not a discovery on my side, that is exactly why I keep saying that the argument `--model_parallel` should be removed. It doesn't actually do anything and is confusing for the user. The call to `model.parallelize()` can always be done outside of `Trainer` IMO, which is why the test can be changed as suggested. We can think of integrating it inside the Trainer later, when the API is stable and actually used, for now I don't see the point of adding this.

> Wrt DataParallel. Why are we turning it on automatically in first place? Why not make it manual and call it --data_parallel

That would be a big breaking change in the API, and beginners actually want to have the parallelism work out of the box when they have several GPUs, so I don't see why change something that works.

stas00 commented:
> The call to model.parallelize() can always be done outside of Trainer IMO, which is why the test can be changed as suggested. 

It doesn't work



>     Wrt DataParallel. Why are we turning it on automatically in first place? Why not make it manual and call it --data_parallel
> 
> That would be a big breaking change in the API, and beginners actually want to have the parallelism work out of the box when they have several GPUs, so I don't see why change something that works.

OK, then the flag should be there with the default On? Surely a user should be able not to run DP and it's not possible at the moment.

stas00 commented:
OK, so I did remove `--model_parallel` - no problem in trainer.py since I used `model.is_parallelizable and model.parallel` - and I now understand that that point is that the user has to activate `model.parallelize()` themselves before passing the `model` to the trainer - i.e. no examples scripts will support MP at the moment.

The problem is `training_args.py` - how do I deal with:

```
        if not self.model_parallel:
            train_batch_size = per_device_batch_size * max(1, self.n_gpu)
        else:
            train_batch_size = per_device_batch_size
```

`self` is args here, and there is no `trainer` object. Suggestions?

But I guess I need to first figure out how to make MP work in trainer at all, I doesn't look it was ever tried or tested. As it fails for me.

stas00 commented:
FWIW, `--model_parallel` works just fine with my Bart PR: https://github.com/huggingface/transformers/pull/9384#issuecomment-756300194 in case someone needs it.

I suspect t5 MP wasn't tested/made to work with `generate` tools (beam search, etc.)

stas00 commented:
OK, I committed the bulk of it, and @sgugger will push some magic to deal with `training_args.py`

stas00 commented:
So now I can see I can jokingly blame this on @sgugger since he wanted it removed all along and so I unconsciously did it during rebasing ;) It's all Freud's fault anyway ;)

stas00 commented:
I added a wrapped first, but it looked out of place so I introduced and documented a new attribute: `self.is_model_parallel` - hope it's loud and clear.

stas00 commented:
@sgugger, I must be doing something wrong - that section of `Important attributes` gets wrapped all funny - so I tried to add bullets and then it gets all messed up, as it bunches it all up into one paragraph.

stas00 commented:
and here is why I  removed `init=False` in https://github.com/huggingface/transformers/pull/9451/commits/a7a39216e99aae60238962ec3d6c96ecf23da42b

The tests were failing with:
```
The tests were failing  TypeError: __init__() got an unexpected keyword argument '_n_gpu'
```
https://circle-production-customer-artifacts.s3.amazonaws.com/picard/forks/5bdabdd888af1f000130874a/278[…]cc8b6d6c390aab800d0cc1350f731a19529ac82f48

stas00 commented:
Thank you for fixing the docs, @sgugger!

LysandreJik commented on src/transformers/trainer.py:
```diff
@@ -267,15 +268,23 @@ def __init__(
                 )
             self.model_init = model_init
 
+        if hasattr(model, "is_parallelizable") and model.is_parallelizable and model.model_parallel:
+            self.is_model_parallel = True
+        else:
+            self.is_model_parallel = False
```
Cool!

LysandreJik reviewed (approved):
Yes, LGTM!
